Content.Shared/Tag/TagSystem.cs

```diff
@@ -15,685 +12,450 @@ public sealed class TagSystem : EntitySystem
-               AddTags(entity, component, ids);
+        return TryAddTags(entityUid, (IEnumerable<ProtoId<TagPrototype>>)tags);
     }
 
     /// <summary>
-    ///     Tries to add the given tags to an entity if it has a
-    ///     <see cref="TagComponent"/> and the tags don't already exist.
+    /// Tries to add the given tags to an entity if it has a
+    /// <see cref="TagComponent"/> and the tags don't already exist.
     /// </summary>
-    /// <param name="entity">The entity to add the tag to.</param>
-    /// <param name="ids">The tags to add.</param>
     /// <returns>
-    ///     true if any tags were added, false otherwise even if they all already existed.
+    /// true if any tags were added, false otherwise even if they all already existed.
     /// </returns>
     /// <exception cref="UnknownPrototypeException">
-    ///     Thrown if one of the ids represents an unregistered <see cref="TagPrototype"/>.
+    /// Thrown if one of the ids represents an unregistered <see cref="TagPrototype"/>.
     /// </exception>
-    public bool TryAddTags(EntityUid entity, IEnumerable<string> ids)
+    public bool TryAddTags(EntityUid entityUid, IEnumerable<ProtoId<TagPrototype>> tags)
     {
-        return _tagQuery.TryComp(entity, out var component) &&
-               AddTags(entity, component, ids);
+        return _tagQuery.TryComp(entityUid, out var component) &&
+               AddTags((entityUid, component), tags);
     }
 
     /// <summary>
-    ///     Checks if a tag has been added to an entity.
+    /// Checks if a tag has been added to an entity.
     /// </summary>
-    /// <param name="entity">The entity to check.</param>
-    /// <param name="id">The tag to check for.</param>
-    /// <returns>true if it exists, false otherwise.</returns>
-    /// <exception cref="UnknownPrototypeException">
-    ///     Thrown if no <see cref="TagPrototype"/> exists with the given id.
-    /// </exception>
-    public bool HasTag(EntityUid entity, string id)
-    {
-        return _tagQuery.TryComp(entity, out var component) &&
-               HasTag(component, id);
-    }
-
-    /// <summary>
-    ///     Checks if a tag has been added to an entity.
-    /// </summary>
-    [Obsolete]
-    public bool HasTag(EntityUid entity, string id, EntityQuery<TagComponent> tagQuery)
-    {
-        return tagQuery.TryGetComponent(entity, out var component) &&
-               HasTag(component, id);
-    }
-
-    /// <summary>
-    ///     Checks if all of the given tags have been added to an entity.
-    /// </summary>
-    /// <param name="entity">The entity to check.</param>
-    /// <param name="id">The tags to check for.</param>
-    /// <returns>true if they all exist, false otherwise.</returns>
-    /// <exception cref="UnknownPrototypeException">
-    ///     Thrown if one of the ids represents an unregistered <see cref="TagPrototype"/>.
-    /// </exception>
-    public bool HasAllTags(EntityUid entity, string id) => HasTag(entity, id);
```

Comment:
Does casting a class to the IEnumerable type create anything? Almost any list implements an iteration method of this interface, and when we cast an instance to the IEnumerable interface, we limit the ability to interact with the class to only the methods and properties described in IEnumerable. I don't think the type conversion operation consumes a lot of resources.

Reply:
Ok, I seem to get the point, but what prevents us from using HasTag, because we still input a collection, and it simply will not cause an overload even if it has one element?

Reply:
> Ok, I seem to get the point, but what prevents us from using HasTag, because we still input a collection, and it simply will not cause an overload even if it has one element?

I'd be fine with just using hasTag if `HasTags` were given a debug assert that checks that people don't accidentally use it when passing in only one tag (which they were doing, which is why the override exists). But I don't think theres a way to differentiate between that and someone just passing in an actual array that happens to only have one element.

Having the override doesn't hurt anything, its only a few lines of codes and stops some trivial mistakes from accidentally slowing things down a bit.

> Does casting a class to the IEnumerable type create anything?

Using `IEnumerable` and uneccesarily using `params []` methods allocates and slows things down. IIRC for IEnumerable its because it has to create an enumerator class. For example, running these benchmarks https://github.com/ElectroJr/RobustToolbox/commit/fd62c6d02f4b65e833f40a04af2059c752130fbe yields:

| Method                  | Mean     | Error    | StdDev   | Ratio | Gen0   | Allocated | Alloc Ratio |
|------------------------ |---------:|---------:|---------:|------:|-------:|----------:|------------:|
| HasTagsArray            | 21.39 ns | 0.129 ns | 0.121 ns |  1.00 |      - |         - |          NA |
| HasTagsList             | 21.70 ns | 0.087 ns | 0.081 ns |  1.01 |      - |         - |          NA |
| HasTagsIEnumerableArray | 29.08 ns | 0.099 ns | 0.093 ns |  1.36 | 0.0038 |      32 B |          NA |
| HasTagsIEnumerableList  | 30.13 ns | 0.098 ns | 0.092 ns |  1.41 | 0.0048 |      40 B |          NA

| Method        | Mean     | Error    | StdDev   | Ratio | Gen0   | Allocated | Alloc Ratio |
|-------------- |---------:|---------:|---------:|------:|-------:|----------:|------------:|
| HasTag        | 10.35 ns | 0.043 ns | 0.040 ns |  1.00 |      - |         - |          NA |
| HasTagsParams | 16.23 ns | 0.105 ns | 0.082 ns |  1.57 | 0.0038 |      32 B |          NA |

They're generally 1.3-1.6 times slower and allocate. In most cases it usually doesn't matter and I wouldn't be picky, but I think that tags get used frequently enough by other systems that it would make a not insignificant difference. E.g., there are systems that iterate over many entities and check for some tags, or things like `InRangeUnobstructed()` which does some tag checks. 

It might be premature optimization, but I'd rather just keep the faster methods around. Though they look like bloat & duplicate code, they're not really hard to maintain and just help make things a bit faster.

Reply:
To clarify, this really only matters for the has-tag methods. Add/remove tags don't get used often enough for it to matter